04_basic_struct/hello.c

```diff
@@ -0,0 +1,240 @@
+// SPDX-License-Identifier: GPL-2.0 OR BSD-3-Clause
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/printk.h>
+#include <linux/kobject.h>
+#include <linux/sysfs.h>
+#include <linux/list.h>
+#include <linux/slab.h> //kmalloc
+#include <linux/string.h>
+
+MODULE_AUTHOR("Oleh Yakymenko <oleh.o.yakymenko@globallogic.com");
+MODULE_DESCRIPTION("Linux kernel data structure");
+MODULE_LICENSE("Dual BSD/GPL");
+MODULE_VERSION("0.1");
+
+static LIST_HEAD(hello_list);
+struct hello_node {
+	struct list_head list;
+	char *data;
+};
+
+
+struct hello_node *list_create_node(const char *s, size_t count)
+{
+	struct hello_node *node = kmalloc(sizeof(*node),
+					GFP_KERNEL);
+
+	if (node == NULL)
+		pr_err("Element hasn`t created. Not enaugh memory for list entry\n");
+
+	node->data = kmalloc_array(count, sizeof(char), GFP_KERNEL);
+
+	strcpy(node->data, s);
+
+	return node;
+}
+
+void list_delete_node(struct hello_node *a_node)
+{
+	if (a_node == NULL)
+		return;
+
+	if (a_node->data != NULL)
+		kfree(a_node->data);
+
+	kfree(a_node);
+}
+
+int list_print(struct list_head *print_list, char *buf)
+{
+	const char *arrow = " -> ";
+	const int arrow_size = strlen(arrow);
+	int buf_offset = 0;
+	struct hello_node *current_node = NULL;
+
+	if (print_list == NULL) {
+		pr_err("Cant print NULL list\n");
+		return 0;
+	}
+
+	if (list_empty(print_list) == 1) {
+
+		sprintf(buf+buf_offset, "%s\n",
+			"List empty!\n"
+			"USAGE:\n"
+			"at_<string> - add to tail\n"
+			"ah_<string> - add to head\n"
+			"rt - remove from tail\n"
+			"rh - remove from head");
+		return strlen(buf);
+	}
+
+	sprintf(buf, "%s", "List: ");
+	buf_offset += strlen("List: ");
+
+	list_for_each_entry(current_node, &hello_list, list) {
+		if (buf_offset + arrow_size + 1 > PAGE_SIZE)
+			break;
+
+		sprintf(buf+buf_offset, "%s", current_node->data);
+		buf_offset += strlen(current_node->data)-1;
```

Comment:
For sprintf. If successful, the total number of characters written is returned excluding the null-character appended at the end of the string, otherwise a negative number is returned in case of failure.